### PR TITLE
FEATURE: Cancel export functionality

### LINKED ISSUE #479: Add a `#cancelExport` method
**Please note we are preparing to participate in Google Code-in, and have reserved this issue for GCI participants - but we'd love to have your help with another one! Please check out https://code.publiclab.org to see more.**

# Problem

LDI provides the ability to export multiple images. 
Demo it here: https://publiclab.github.io/Leaflet.DistortableImage/examples/select.html
by 
- `ctrl + click`ing on one or more images. 
- A single toolbar for all of them will appear in the topleft corner. 
- Click the top icon to begin export
- Open your developer console to see updates on the status
- While the export is in progress a spinner icon appears. 

The problem now is that once the export starts - there is no way to cancel it besides reloading the page.

We would like to add this capability.

# Solution:

1) Add a `cancelExport` method to the following file, (where you can also find `startExport`):
https://github.com/publiclab/Leaflet.DistortableImage/blob/978c33915620d241eaee70c224966e7d52205d50/src/edit/DistortableCollection.Edit.js#L194-L257

Note you may need to refactor `startExport` to be able to access and clear the interval early

2) Add a red `X` icon that will appear when you hover over the spinner icon during export. On clicking this icon, you will trigger `cancelExport`.

**More Relevant information:**

- https://github.com/publiclab/Leaflet.DistortableImage/blob/main/src/edit/actions/ExportAction.js - this is the file for the export action. In `addHooks` we
   - `if (edit instanceof L.DistortableImage.Edit) { edit._getExport(); }` Single image instances have a handler class `L.DistortableImage.Edit` that triggers a single image export `edit._getExport();` (you can test this one by clicking on any image and hitting the export icon on its attached toolbar). Multiple images have a separate handler `L.DistortableCollection.Edit` that calls the multiple export function, `edit.startExport()`, which you demo'd earlier. We check which one to call.
  - if its multiple, follows are method that switch the action's icon to the spinner, and add the `loader` class to it to animate it's spin via css. 
  - Then a call to the multiple export function `edit.startExport()`

https://github.com/publiclab/Leaflet.DistortableImage/blob/f5378980b8712bc8b95faa1d9ce26e2fda5d6cd8/src/edit/actions/ExportAction.js#L24-L35
 
- We use SVG icons, with a simple system for rendering the icons in JS dynamically. Usage can be found here:  https://github.com/publiclab/Leaflet.DistortableImage/wiki/SVG-Icon-System.

Heres the svg code for the recommended asset: 
```SVG
<svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 18 18">
  <polygon points="13.279 5.779 12.221 4.721 9 7.942 5.779 4.721 4.721 5.779 7.942 9 4.721 12.221 5.779 13.279 9 10.057 12.221 13.279 13.279 12.221 10.057 9"/>
</svg>
```

Just create a file in `assets/svg` called `cancel.svg` and paste this code. Then, follow the instructions in the linked wiki.

- Checking out this issue might help too - it provides a little bit of additional information https://github.com/publiclab/Leaflet.DistortableImage/issues/470

##  💬   Get help
If you need any help - here are some options: 
- Comment below
- Join our gitter chat at https://gitter.im/publiclab/publiclab

Labeled `hard` because requires a deeper understanding of LDI's moving parts. Here to help with any questions!!

### PR DESCRIPTION
**Pull request by VladimirMikulic**
[Demo](https://streamable.com/r5cxv)

Resolves #479 

@sashadev-sky @jywarren @SidharthBansal  @ananyaarun could you review this, please?

### CODE REVIEW

File: assets/icons/svg/cancel.svg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 18 18">
+  <polygon points="13.279 5.779 12.221 4.721 9 7.942 5.779 4.721 4.721 5.779 7.942 9 4.721 12.221 5.779 13.279 9 10.057 12.221 13.279 13.279 12.221 10.057 9"/>
+</svg>
```
Comment (sashadev-sky) [inline assets/icons/svg/cancel.svg]: if we could add a new line at the end of this file that would be great!
```suggestion
</svg>

```

File: dist/leaflet.distortableimage.js [hunk 2]
```diff
@@ -1391,6 +1391,7 @@ L.KeymapperIconSet = L.IconSet.extend({
 /* eslint-disable max-len */
 L.ToolbarIconSet = L.IconSet.extend({
   _symbols:
+    '<symbol viewBox="0 0 18 18" id="cancel"><path d="M13.279 5.779l-1.058-1.058L9 7.942 5.779 4.721 4.721 5.779 7.942 9l-3.221 3.221 1.058 1.058L9 10.057l3.221 3.222 1.058-1.058L10.057 9z"/></symbol>' +
```
Comment (sashadev-sky) [inline dist/leaflet.distortableimage.js]: Looks like we keep these in alphabetical order so please move `cancel` after `border_outer`

File: dist/leaflet.distortableimage.js [hunk 3]
```diff
@@ -1629,28 +1635,99 @@ L.ExportAction = L.EditAction.extend({
     options.toolbarIcon = {
       svg: true,
       html: 'get_app',
-      tooltip: tooltip,
+      tooltip: tooltip
```
Comment (sashadev-sky) [inline dist/leaflet.distortableimage.js]: we use dangling commas for objects
```suggestion
      tooltip: tooltip,
```

File: dist/leaflet.distortableimage.js (hunk 2 above)
Comment (VladimirMikulic) [reply dist/leaflet.distortableimage.js]: :+1:

File: dist/leaflet.distortableimage.js (hunk 3 above)
Comment (VladimirMikulic) [reply dist/leaflet.distortableimage.js]: :+1:

**Comment by VladimirMikulic** [comment]
Hi @sashadev-sky, I am glad you are back. I've done the requested changes.

**Comment by sashadev-sky** [comment]
@VladimirMikulic you too!! I am going to review right now

**Comment by VladimirMikulic** [comment]
@sashadev-sky thank you for the review. I've implemented the changes that you've requested.

**Comment by sashadev-sky** [comment]
also wont let me add you as a reviewer for some reason but here it is #519 (you're not obligated to review it only if you want)

**Comment by VladimirMikulic** [comment]
@sashadev-sky it's fine to close this and specify me as an author. Just to make sure, my points are awarded for this task, right?

**Comment by sashadev-sky** [comment]
Yes great job here!! I'll keep it open until we merge the other one

**Comment by SidharthBansal** [comment]
In case of any conflicts regarding GCI you all can directly contact me. I
am present to resolve all the conflicts.
Thanks

**Comment by SidharthBansal** [comment]
Great work going on.

**Comment by VladimirMikulic** [comment]
@SidharthBansal thank you very much, this PR won't get merged directly. Instead, @jywarren and @sashadev-sky have copied these changes in #519 so the points still count.

**Comment by SidharthBansal** [comment]
If that code was yours, then get the task approved on GCI dashboard asap.
Program is finishing up soon. So, I request you to get points for all the tasks which you completed before the deadline. Also get the already approved task's related PR merged. We will NOT consider points for tasks which are not MERGED**.


** We will however review it again. If we will find that we should give marks for that unmerged task. We will give you points. We will try to be as fair as possible to both you and other participants.

**Comment by SidharthBansal** [comment]
In case of doubts, catch me anywhere 🚀

**Comment by VladimirMikulic** [comment]
Thanks, @SidharthBansal. As you can see #519 is merged and this PRs code is in it.

**Comment by SidharthBansal** [comment]
Great! Please grab points then :tada:

**Comment by SidharthBansal** [comment]
I mean get approval at GCI dashboard and close this pr if it is not needed. Thanks

**Comment by VladimirMikulic** [comment]
The task has been approved already. @sashadev-sky I am closing this.

**Pull request closed without merging**